pkg/crypto/certloader/watcher.go

```diff
@@ -0,0 +1,239 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certloader
+
+import (
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// Watcher is a set of TLS configuration files including CA files, and a
+// certificate along with its private key. The files are watched for change and
+// reloaded automatically.
+type Watcher struct {
+	*FileReloader
+	log       logrus.FieldLogger
+	fswatcher *fsnotify.Watcher
+	stop      chan struct{}
+}
+
+// NewWatcher returns a Watcher that watch over the given file
+// paths. The given files are expected to already exists when this function is
+// called. On success, the returned Watcher is ready to use.
+func NewWatcher(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (*Watcher, error) {
+	r, err := NewFileReloaderReady(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	// An error here would be unexpected as we were able to create a
+	// FileReloader having read the files, so the files should exist and be
+	// "watchable".
+	fswatcher, err := newFsWatcher(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	w := &Watcher{
+		FileReloader: r,
+		log:          log,
+		fswatcher:    fswatcher,
+		stop:         make(chan struct{}),
+	}
+
+	w.Watch()
+	return w, nil
+}
+
+// EventualWatcher returns a channel where exactly one Watcher will be sent
+// once the given files are ready and loaded. This can be useful when the file
+// paths are well-known, but the files themselves don't exist yet. Note that
+// the requirement is that the file directories must exists.
+func EventualWatcher(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (<-chan *Watcher, error) {
+	r, err := NewFileReloader(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	fswatcher, err := newFsWatcher(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	w := &Watcher{
+		FileReloader: r,
+		log:          log,
+		fswatcher:    fswatcher,
+		stop:         make(chan struct{}),
+	}
+
+	res := make(chan *Watcher)
+	go func(res chan<- *Watcher) {
+		defer close(res)
+		// Attempt a reload without having received any fs notification in case
+		// all the files are already there. Note that the keypair and CA are
+		// reloaded separately as a "partial update" is still useful: If the
+		// FileReloader is "half-ready" (e.g. has loaded the keypair but failed
+		// to load the CA), we only need a successfully handled CA related fs
+		// notify event to become Ready (in other words, we don't need to
+		// receive a fs event for the keypair in that case to become ready).
+		_, keypairErr := w.ReloadKeypair()
+		_, caErr := w.ReloadCertificateAuthority()
+		ready := w.Watch()
+		if keypairErr == nil && caErr == nil {
+			log.Debug("ready")
```

Comment:
Is this log message sparse on purpose or simply a leftover?

Reply:
Tried to improve it and now we get

```
level=debug msg="TLS configuration ready" config=tls-server subsys=hubble
```